### PR TITLE
Fix `save_hyperparameters` ignore precedence in subclasses

### LINKED ISSUE #21488: `save_hyperparameters(ignore=...)` is not persistent across inheritance; ignored params reappear when base class also calls `save_hyperparameters`
### Bug description

When `save_hyperparameters()` is called in both a superclass and a subclass, using `ignore=...` in the subclass does **not reliably remove ignored parameters**.
If the base class also calls `save_hyperparameters()`, ignored parameters can be re-added later and end up in  `model.hparams` and checkpoints

This breaks expected inheritance semantics.

This is a issue where there is a common Lightning base class for many models ( or data modules) and call `save_hyperparameters()` in the base class to avoid duplication. We then have multiple subclasses, and for some specific subclasses we need to ignore a certain of parameters and are not needed in checkpoint too.

### What version are you seeing the problem on?

master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
class BaseModel(LightningModule):
    def __init__(self, arg1, arg2):
        super().__init__()
        self.save_hyperparameters()

class ChildModel(BaseModel):
    def __init__(self, arg1, arg2):
        super().__init__(arg1, arg2)
        self.save_hyperparameters(ignore="arg2")

model = ChildModel(arg1=1, arg2=2)

assert "arg2" not in model.hparams
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

This happens because `ignore` only applies to a **single call** and later calls to `save_hyperparameters()` overwrite/merge without remembering ignored keys

## **Potential Fix**

Make `ignore` **persistent per instance**:

1. Track ignored hyperparameters on the object (e.g. `self._ignore_hparams: set[str]`)
2. Merge ignores across calls before called `save_hyperparameters` **internal function** in https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/mixins/hparams_mixin.py#L131

```python
if not hasattr(self, "_ignore_hparams"):
    self._ignore_hparams = set()

if ignore is not None:
    self._ignore_hparams.update(ignore if isinstance(ignore, (list, tuple)) else [ignore])

# Always pass the cumulative ignore set to save_hyperparameters
ignore = list(self._ignore_hparams)

...
save_hyperparameters(self, *args, ignore=ignore, frame=frame, given_hparams=given_hparams)
```

### Happy to raise a PR to fix this if this is acknowledged as bug

cc @ethanwharris @justusschock @lantiga

### PR DESCRIPTION
**Pull request by aditya0by0**
## What does this PR do?

Enable hyperparameters to be ignored when save_hyperparameters is called in both a base class and a subclass, with subclass ignore rules overriding those from the base class.

Fixes #21488
closes #21489

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21490.org.readthedocs.build/en/21490/

### CODE REVIEW

File: tests/tests_pytorch/models/test_hparams.py [hunk 1]
```diff
@@ -838,7 +838,9 @@ def __init__(self, arg1, arg2, arg3):
 
     # test proper property assignments
     assert model.hparams.arg1 == 14
-    for arg in ignore:
+
+    ignore_args = ignore if isinstance(ignore, (list, tuple)) else [ignore]
+    for arg in ignore_args:
```
Comment (aditya0by0) [inline tests/tests_pytorch/models/test_hparams.py:843]: The reason for this change is when ignore="arg2", the loop was iterating over the characters of the string "arg2".

**Review by deependujha — CHANGES_REQUESTED** [review]
Let’s agree on the approach first, then proceed with a single, coherent implementation.

check [comment](https://github.com/Lightning-AI/pytorch-lightning/issues/21488#issuecomment-3756592667) and #21490 #21489

**Review by deependujha — APPROVED** [review]

File: tests/tests_pytorch/models/test_hparams.py [hunk 2]
```diff
@@ -854,7 +856,34 @@ def __init__(self, arg1, arg2, arg3):
...
+            super().__init__(arg1=arg1, arg2=arg2, arg3=arg3)
+            self.save_hyperparameters(ignore=ignore)
+
+    model = LocalModel(arg1=14, arg2=90, arg3=50)
+
+    # `arg1` was ignored by the base class,
+    # but the subclass did not ignore it, so it should be present.
+    assert model.hparams.arg1 == 14
+    assert model.hparams.arg4 == 100
+    assert model.hparams.arg5 == 66
```
Comment (deependujha) [inline tests/tests_pytorch/models/test_hparams.py]: Why should `arg5` be available?

It was passed as an hyperparameter to base class, but it ignored it.

Derived class never received it as an hyperparameter, so why will it save it.

Comment (aditya0by0) [reply tests/tests_pytorch/models/test_hparams.py]: My mental model for this the change should be equivalent to the below code. The below test passes successfully. The behaviour should be as if we have just called `self.save_hyperparameters()` in base class without any ignore parameter.


```python
    class BaseBoringModel(BoringModel):
        def __init__(self, arg1, arg2, arg3, arg4=100, arg5=66):
            super().__init__()
            self.save_hyperparameters()

    class LocalModel(BaseBoringModel):
        def __init__(self, arg1, arg2, arg3):
            super().__init__(arg1=arg1, arg2=arg2, arg3=arg3)
            self.save_hyperparameters(ignore="arg2")

    model = LocalModel(arg1=14, arg2=90, arg3=50)
    assert model.hparams.arg1 == 14
    assert model.hparams.arg4 == 100
    assert model.hparams.arg5 == 66
    assert "arg2" not in model.hparams

    trainer = Trainer(default_root_dir=tmp_path, max_epochs=1, overfit_batches=0.5)
    trainer.fit(model)

    raw_checkpoint_path = _raw_checkpoint_path(trainer)
    raw_checkpoint = torch.load(raw_checkpoint_path, weights_only=True)
    assert LightningModule.CHECKPOINT_HYPER_PARAMS_KEY in raw_checkpoint
    assert raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]["arg1"] == 14
    assert raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]["arg5"] == 66
    assert "arg2" not in raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]
```

Comment (deependujha) [reply tests/tests_pytorch/models/test_hparams.py]: not sure what you meant. This behaviour should remain unchanged in this PR.

When you call `self.save_hyperparameters()`, it inspect the stacktrace and gets the args. Calling `self.save_hyperparameters()` in derived class, will have last stacktrace as `derived class's __init__`, not `base class __init__`.

I'm not sure, if i'm understanding you correctly. Please be more verbose.😅

**Comment by aditya0by0** [comment]
@deependujha, sorry if my previous comments has cause any confusion. Let me try to explain this more clearly step by step.

The agreed-upon and expected behavior is:

> **When `save_hyperparameters` is called in both a base class and a child class, the child class call should take absolute precedence.**
> Any `ignore` rules specified in the base class should be completely overridden and should not affect the final `hparams` or checkpoint state.

I will try to explain my point of view using below two illustrative cases.


## Case 1: Base class has no ignore rules

In this scenario, `save_hyperparameters` is called in the base class without any ignore rules, and the child class ignores only `arg2`.

```python
class Base(BoringModel):
    def __init__(self, arg1, arg2, arg3, arg4=100, arg5=66):
        super().__init__()
        self.save_hyperparameters()

class Child(Base):
    def __init__(self, arg1, arg2, arg3):
        super().__init__(arg1=arg1, arg2=arg2, arg3=arg3)
        self.save_hyperparameters(ignore="arg2")

model = Child(arg1=14, arg2=90, arg3=50)
```

### Expected and observed behavior (✅ passes)

* `arg1`, `arg3`, `arg4`, and `arg5` are present in `model.hparams`
* `arg2` is correctly ignored
* `arg4` and `arg5`, although not explicitly passed to the child class, are still saved as part of the model state
* The same behavior is reflected in the checkpoint

Assertions:

```python
assert model.hparams.arg1 == 14
assert model.hparams.arg3 == 50
assert model.hparams.arg4 == 100
assert model.hparams.arg5 == 66
assert "arg2" not in model.hparams

trainer = Trainer(default_root_dir=tmp_path, max_epochs=1, overfit_batches=0.5)
trainer.fit(model)
raw_checkpoint_path = _raw_checkpoint_path(trainer)
raw_checkpoint = torch.load(raw_checkpoint_path, weights_only=True)
model_hparams = raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]


assert model_hparams["arg1"] == 14
assert model_hparams["arg3"] == 50
assert model_hparams["arg4"] == 100
assert model_hparams["arg5"] == 66
assert "arg2" not in model_hparams
```

This behavior is correct and expected.


## Case 2: Base class defines ignore rules, child class defines its own

Now we modify the base class to ignore `arg1` and `arg5`.
The child class still ignores only `arg2`.

```python
class Base(BoringModel):
    def __init__(self, arg1, arg2, arg3, arg4=100, arg5=66):
        super().__init__()
        self.save_hyperparameters(ignore=("arg1", "arg5"))  # ignore rules should not be taken into account as overridden by base class

class Child(Base):
    def __init__(self, arg1, arg2, arg3):
        super().__init__(arg1=arg1, arg2=arg2, arg3=arg3)
        self.save_hyperparameters(ignore="arg2")  # only this rule should apply

model = Child(arg1=14, arg2=90, arg3=50)
```

### Expected behavior

Since the child class calls `save_hyperparameters`, **its ignore rules should fully override the base class rules**.
Therefore, the behavior should be identical to Case 1:

* `arg1`, `arg3`, `arg4`, and `arg5` should be present
* `arg2` should be ignored
* Base class ignore rules (`arg1`, `arg5`) should have no effect



#### But `arg5` is missing from `hparams` and the checkpoint in this case

In this scenario, `arg5` is **not** present in either `model.hparams` or the checkpoint (i.e., `'AttributeDict' object has no attribute 'arg5'`). This is unexpected.

Notably:

* `arg4` was **never passed** as a hyperparameter to the derived class, yet it is still present in `hparams`
* By the same logic, `arg5` which is also a default argument of the base class and part of the model state **should also be present**

Since the derived class calls `save_hyperparameters()` and only ignores `arg2`, there is no clear reason for `arg5` to be excluded while `arg4` is retained.

Assertions:

```python
assert model.hparams.arg1 == 14
assert model.hparams.arg3 == 50
assert model.hparams.arg4 == 100
assert model.hparams.arg5 == 66 # FAILS
assert "arg2" not in model.hparams

trainer = Trainer(default_root_dir=tmp_path, max_epochs=1, overfit_batches=0.5)
trainer.fit(model)
raw_checkpoint_path = _raw_checkpoint_path(trainer)
raw_checkpoint = torch.load(raw_checkpoint_path, weights_only=True)
model_hparams = raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]

assert model_hparams["arg1"] == 14
assert model_hparams["arg3"] == 50
assert model_hparams["arg4"] == 100
assert model_hparams["arg5"] == 66 #FAILS
assert "arg2" not in model_hparams
```

**Comment by deependujha** [comment]
Hi @aditya0by0, is there any real usecase for this?

Also as stated earlier: When you call `self.save_hyperparameters()`, it inspect the stacktrace and gets the args. Calling `self.save_hyperparameters()` in derived class, will have last stacktrace as `derived class's __init__`, not `base class __init__`.

If base class ignored it, and you want to overwrite it in derived class, add it as a hyperparameter in derived class as well.

```python
class Base(BoringModel):
    def __init__(self, arg1, arg2, arg3, arg4=100, arg5=66):
        super().__init__()
        self.save_hyperparameters(ignore=("arg1", "arg5"))  # ignore rules should not be taken into account as overridden by base class

class Child(Base):
    def __init__(self, arg1, arg2, arg3, arg5=66): # also add arg5 if you want to force it to be saved.
        super().__init__(arg1=arg1, arg2=arg2, arg3=arg3)
        self.save_hyperparameters(ignore="arg2")  # only this rule should apply

model = Child(arg1=14, arg2=90, arg3=50)
```

---

This implementation makes sense as it doesn't overcomplicate for a simple usecase.

**Comment by aditya0by0** [comment]
Thanks for the clarification, completely agree with your point.

**Review by tchaton — APPROVED** [review]
